src/audio/pipeline.c

```diff
@@ -761,6 +762,68 @@ static int pipeline_trigger_on_core(struct pipeline *p, struct comp_dev *host,
 	return ret;
 }
 
+/*
+ * check if we need real trigger for the pipeline.
+ * return true if we need process real trigger, false if doesn't.
+ */
+static bool pipeline_is_trigger_needed(struct pipeline *p, int cmd)
+{
+	int ret;
+
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+	case COMP_TRIGGER_RELEASE:
+		/* for xrun, we need prepare pipeline before start it */
+		if (p->status == COMP_STATE_PREPARE && p->xrun_bytes) {
```

Comment:
This is doing more that "is_trigger_needed". Looks like this is logic for handling XRUNs

Reply:
Yes, I try doing some calibration here to make our FW more robust/smart(e.g. we get trigger start without ever prepared, SOF driver don't send prepare IPC actually).